pkg/spade-test/spade-test.go

```diff
@@ -0,0 +1,284 @@
+	ipld "github.com/ipld/go-ipld-prime"
+	_ "github.com/ipld/go-ipld-prime/codec/dagcbor"
+	_ "github.com/ipld/go-ipld-prime/codec/dagjson"
+	_ "github.com/ipld/go-ipld-prime/codec/raw"
+)
+
+var logger = logging.Logger("spade-test")
+
+type SingleContentRouter struct {
+	AddrInfo peer.AddrInfo
+}
+
+func (s SingleContentRouter) PutValue(context.Context, string, []byte, ...routing.Option) error {
+	return routing.ErrNotSupported
+}
+
+func (s SingleContentRouter) GetValue(context.Context, string, ...routing.Option) ([]byte, error) {
+	return nil, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) SearchValue(ctx context.Context, key string, opts ...routing.Option) (
+	<-chan []byte, error) {
+	return nil, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) Provide(context.Context, goCid.Cid, bool) error {
+	return routing.ErrNotSupported
+}
+
+func (s SingleContentRouter) FindProvidersAsync(context.Context, goCid.Cid, int) <-chan peer.AddrInfo {
+	ch := make(chan peer.AddrInfo)
+	go func() {
+		ch <- s.AddrInfo
+		close(ch)
+	}()
+	return ch
+}
+
+func (s SingleContentRouter) FindPeer(context.Context, peer.ID) (peer.AddrInfo, error) {
+	return peer.AddrInfo{}, routing.ErrNotFound
+}
+
+func (s SingleContentRouter) Bootstrap(context.Context) error {
+	return nil
+}
+
+func (s SingleContentRouter) Close() error {
+	return nil
+}
+
+type MessageReceiver struct {
+	BSClient       *bsclient.Client
+	MessageHandler func(ctx context.Context, sender peer.ID, incoming bsmsg.BitSwapMessage)
+}
+
+func (m MessageReceiver) ReceiveMessage(
+	ctx context.Context,
+	sender peer.ID,
+	incoming bsmsg.BitSwapMessage) {
+	m.BSClient.ReceiveMessage(ctx, sender, incoming)
+	m.MessageHandler(ctx, sender, incoming)
+}
+
+func (m MessageReceiver) ReceiveError(err error) {
+	m.BSClient.ReceiveError(err)
+}
+
+func (m MessageReceiver) PeerConnected(id peer.ID) {
+	m.BSClient.PeerConnected(id)
+}
+func (m MessageReceiver) PeerDisconnected(id peer.ID) {
+	m.BSClient.PeerDisconnected(id)
+}
+
+type BitswapClient struct {
+	host    host.Host
+	timeout time.Duration
+}
+
+func NewBitswapClient(host host.Host, timeout time.Duration) BitswapClient {
+	return BitswapClient{
+		host:    host,
+		timeout: timeout,
+	}
+}
+
+// Attempts to decode the block data into a node and return its links
+func FindLinks(ctx context.Context, blk blocks.Block) ([]datamodel.Link, error) {
+	if blk.Cid().Prefix().Codec == goCid.Raw {
+		// This can happen at the bottom of the tree
+		return nil, errors.New("raw block encountered " + blk.Cid().String())
+	}
+
+	decoder, err := cidlink.DefaultLinkSystem().DecoderChooser(cidlink.Link{Cid: blk.Cid()})
+
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := ipld.Decode(blk.RawData(), decoder)
+	if err != nil {
+		return nil, err
+	}
+
+	links, err := traversal.SelectLinks(node)
+	if err != nil {
+		return nil, err
+	}
+
+	return links, nil
+}
+
+// Returns the raw block data, the links, and error if any
+func (c BitswapClient) RetrieveBlock(
```

Comment:
basically the `bitswap.go` `Retrieve` function but reworked to just return the block at the end of it (simplified for this tech spike by removing the `task` returns)

Reply:
@xinaxu for the integration, do you think we should upgrade the existing `Retrieve` func to work like this, passing back the `blk` at the end, and then have the calling code determine what to do next? This way, we could have a single `bitswap Retrieve()` function that both the Spade and "regular" retrieval tests use. Might this also be useful for V2 payload verification stuff?